snapcraft/file_utils.py

```diff
@@ -150,18 +150,22 @@ def _search_and_replace_contents(file_path, search_pattern, replacement):
     if os.path.islink(file_path):
         return
 
-    with open(file_path, 'r+') as f:
-        try:
-            original = f.read()
-        except UnicodeDecodeError:
-            # This was probably a binary file. Skip it.
-            return
-
-        replaced = search_pattern.sub(replacement, original)
-        if replaced != original:
-            f.seek(0)
-            f.truncate()
-            f.write(replaced)
+    try:
+        with open(file_path, 'r+') as f:
+            try:
+                original = f.read()
+            except UnicodeDecodeError:
+                # This was probably a binary file. Skip it.
+                return
+
+            replaced = search_pattern.sub(replacement, original)
+            if replaced != original:
+                f.seek(0)
+                f.truncate()
+                f.write(replaced)
+    except PermissionError as e:
+        logger.debug('Unable to open {path} for writing: {error}'.format(
+            path=file_path, error=e))
```

Comment:
That's right - the throw PermissionError happens as soon as `open` is called with `r+`.

As for the large try-catch block, I believe that's the most pythonic way to handle this situation (asking forgiveness instead of permission). From what I've read, the try-catch will have little-to-no effect on performance unless the catch block is hit frequently, which shouldn't be the the case here (has only affected me with one or two files over the course of quite a few snap buidings).